### DISCUSSION TITLE
History on the Pull requests

### DISCUSSION THREAD

**Discussion by PeterBocan** [community/community#132528] [Apps, API and Webhooks]
### Select Topic Area

General

### Body

Hello, 

I am curious why isn't there `history` on PRs like there is the option on Commits? I am trying to filter out PRs in a specific time range, but I am not sure how to achieve it in current API.

**Comment by ryantusi** [comment]
Hey Peter Bocan

Yes, GitHub maintains a history of Pull Requests (PRs) as well. Each PR contains detailed information about the changes made, discussions, reviews, and merge status. I'm not sure what API you are talking about though. Here's how you can access and view the history of a Pull Request on GitHub:

1. **Navigate to the Repository:** Open the GitHub repository where the PR is located.
2. **Go to the Pull Requests Tab:** Click on the `Pull requests` tab at the top of the repository page.
3. **Open a Specific PR:** In the list of PRs, click on the title of the PR you want to view. This will open the PR's main page.

Once you're on the PR page, you can explore various sections that provide historical information and if you want to view detailed PR History::

1. **Conversation Tab:** This tab shows the overall conversation and discussion around the PR. It includes comments, code reviews, and any status checks or CI/CD integrations that have run.

2. **Commits Tab:** This tab lists all the commits included in the PR. You can click on each commit to see the changes made and the commit message.

3. **Files Changed Tab:** This tab shows a diff of the changes introduced by the PR. It lists all files that were changed and highlights the specific changes within each file.

4. **Checks Tab:** This tab shows the status of any checks that were run on the PR, such as automated tests or CI/CD pipelines.

5. **Timeline:** On the right side of the Conversation tab, you’ll see a timeline that includes events like when the PR was opened, comments made, reviews requested, reviews submitted, and when the PR was merged or closed.

You can also do these:
- **Review Comments:** Inline comments can be added directly to the code in the `Files changed` tab, allowing for specific discussions on code lines.

- **Review Status:** The status of reviews (_approved, changes requested, or comments_) is visible in the Conversation tab and timeline.

- **Merge Information:** If the PR is merged, the timeline will show when and by whom it was merged. It also shows if the PR was closed without merging.

- **Linked Issues:** PRs can be linked to issues. Any linked issues will be shown in the timeline, and closing the PR can automatically close the linked issues if the right keywords are used (e.g., `closes #issue_number`).

Well if you want a step-by-step guide to access the history of a specific PR, I can provide tutorial using my github repository:

1. `Go to any of my Repository`
2. `Pull Requests Tab`
3. `Select a PR`
4. `Explore History`

Hope I made it simple enough for you to understand. Let me know if you still find trouble with it.
**_Happy Coding and make sure to mark it answered if you feel so, plus let's connect._**

**Comment by PeterBocan** [comment]
If I want to fetch historical commits using GraphQL I will do it the following way:

```
query FetchCommits {
  organization(login: "xyz") {
    repositories(first: 20 isFork: false isArchived: false)  {
      totalCount
      pageInfo{hasNextPage}
      nodes {
        ... on Repository {
          name
          defaultBranchRef {
            name 
            target {
              ... on Commit {
                history(since:"2024-07-01T00:00:00"  until:"2024-07-11T14:00:00") {
                  totalCount 
                  nodes {
                    ... on Commit {
                       message
                       messageBody
                       associatedPullRequests {
                         nodes {
                           bodyText
                         }
                       }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

note the `history(...)` above. There is no such thing on the PRs, e.g. 

```
          defaultBranchRef {
            name 
            target {
              ... on PullRequests {
                history(since:"2024-07-01T00:00:00"  until:"2024-07-11T14:00:00") {
                  totalCount 
                  nodes {
                    ... on PullRequest {
                       message
                       messageBody
                       associatedPullRequests {
                         nodes {
                           bodyText
                         }
                       }
                    }
                  }
                }
              }
            }
          }
```

or is there?

**Comment by PeterBocan** [comment]
Can I ask maybe for a `PullRequestHistoryConnection` which would mirror the [CommitHistoryConnection](https://docs.github.com/en/graphql/reference/objects#commithistoryconnection) ?

**Comment by ryantusi** [comment]
Oh the GraphQL API, there isn't a direct `history` field on `PullRequest` objects like there is for `Commit` objects. The `history` field is specific to the `Commit` object, which allows you to traverse the commit history of a branch.

To fetch pull requests within a specific date range, you need to use a different approach. Instead of using a `history` field, you can filter pull requests based on their `createdAt` or `updatedAt` timestamps.  Do it like this:
1. **Querying Pull Requests by Date Range**: You can fetch pull requests for a repository and filter them using the `createdAt` field. 
2. **Pagination**: Make sure to handle pagination if you expect a large number of pull requests.

Here's an example query to fetch pull requests created between `2024-07-01T00:00:00` and `2024-07-11T14:00:00`:

```
query FetchPullRequests {
  organization(login: "xyz") {
    repositories(first: 20, isFork: false, isArchived: false) {
      totalCount
      pageInfo { hasNextPage }
      nodes {
        name
        pullRequests(first: 100, states: [OPEN, CLOSED, MERGED], orderBy: {field: CREATED_AT, direction: DESC}) {
          nodes {
            title
            bodyText
            createdAt
            updatedAt
            author {
              login
            }
          }
        }
      }
    }
  }
}
```

This query fetches the first 100 pull requests from each repository, ordered by their creation date in descending order. You can adjust the `first` argument to fetch more or fewer pull requests as needed.

To filter the pull requests by date range on the client side, you can use the `createdAt` field returned in the response. If you need to fetch more pull requests, handle pagination using the `pageInfo` field.

Also keep in mind that this approach retrieves pull requests within the given repositories but does not directly support filtering by date range within the GraphQL API query itself. You will need to perform the filtering logic on the client side after retrieving the pull requests. Just start thinking and using different logic if you can't get it in one way. There is always a solution to everything, if not, congratulations, you discovered something.

Hope this helped answer your question.

**Comment by PeterBocan** [comment]
I am mostly interested in the closed PRs. Filtering out on the client side, though possible, it feels a bit lacklusterish.

**Comment by ryantusi** [comment]
Yeah you just adjust the states to ["CLOSED"] if you are mostly interested in the closed PRs. Probably lacklusterish, but the only way I can find.
